packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -568,13 +592,77 @@ Hardhat Network's forking functionality only works with blocks from at least spu
   ): Promise<EstimateGasResult> {
     // We get the CallParams and transform it into a TransactionParams to be
     // able to run it.
-    const txParams = {
-      ...callParams,
-      nonce: await this._getNonce(
-        new Address(callParams.from),
-        blockNumberOrPending
-      ),
-    };
+    const nonce = await this._getNonce(
+      new Address(callParams.from),
+      blockNumberOrPending
+    );
+
+    // TODO: This is more complex in Geth, we should make sure we aren't missing
+    //  anything here.
+
+    let gasPrice = callParams.gasPrice;
+    let maxFeePerGas = callParams.maxFeePerGas;
+    let maxPriorityFeePerGas = callParams.maxPriorityFeePerGas;
+
+    if (!this.isEip1559Active()) {
+      if (gasPrice === undefined) {
+        gasPrice = await this.getGasPrice();
+      }
+    } else if (gasPrice === undefined) {
+      if (maxPriorityFeePerGas === undefined) {
+        maxPriorityFeePerGas = await this.getMaxPriorityFeePerGas();
+
+        if (
+          callParams.maxFeePerGas !== undefined &&
+          callParams.maxFeePerGas.lt(maxPriorityFeePerGas)
+        ) {
+          maxPriorityFeePerGas = callParams.maxFeePerGas;
+        }
+      }
+
+      if (callParams.maxPriorityFeePerGas === undefined) {
+        if (blockNumberOrPending !== "pending") {
+          const block = await this.getBlockByNumber(blockNumberOrPending);
+
+          maxFeePerGas = maxPriorityFeePerGas.add(
+            block?.header.baseFeePerGas ?? new BN(0)
+          );
+        }
+
+        const baseFeePerGas = await this.getNextBlockBaseFeePerGas();
+        maxFeePerGas = baseFeePerGas!.muln(2).add(maxPriorityFeePerGas);
```

Comment:
I think these two statements should be inside an `else` block, otherwise the previous code is useless, isn't it? And if that's the case, I would switch the blocks (that is, I would start with the `=== "pending"` case, to avoid a conceptual double-negative)